### PR TITLE
Update ippool controller

### PR DESCRIPTION
**Pull request by zhengxiexie**
Update type of ippool to optional.
Overwrite DefaultSubnetAccessMode of vpcnetworkconfig to ippool's type.
Get private or public IpBlockPath from VPCInfo.
Update Type public/private to Public/Private.
Try next ipblock if the previous are exhausted
	If the error message contains `not have spare capacity to satisfy`,
	it shows the ip block has no more capacity, then try next ip block
	from the list of PrivateIpv4Blocks or ExternalIpv4Blocks
	to build IpBlockPath.

### CODE REVIEW

File: pkg/controllers/ippool/ippool_controller.go [hunk 1]
```diff
@@ -134,6 +132,16 @@ func (r *IPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 			log.V(1).Info("added finalizer on ippool CR", "ippool", req.NamespacedName)
 		}
 
+		if obj.Spec.Type == "" {
```
Comment (dantingl) [inline pkg/controllers/ippool/ippool_controller.go:135]: prefixLength should also be got from default subnet size?

Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:135]: > prefixLength should also be got from default subnet size?
Actutally, PrefixLength is defined in SubnetRequest, it is specified by WCP.

File: pkg/controllers/ippool/ippool_controller_test.go [hunk 2]
```diff
@@ -186,6 +186,7 @@ func TestIPPoolReconciler_Reconcile(t *testing.T) {
 	k8sClient.EXPECT().Get(ctx, gomock.Any(), sp).Return(nil).Do(func(_ context.Context, _ client.ObjectKey, obj client.Object, option ...client.GetOption) error {
 		v1sp := obj.(*v1alpha2.IPPool)
 		v1sp.ObjectMeta.DeletionTimestamp = nil
+		v1sp.Spec.Type = "public"
```
Comment (dantingl) [inline pkg/controllers/ippool/ippool_controller_test.go:189]: Public

File: pkg/nsx/services/ippool/builder.go [hunk 3]
```diff
@@ -92,15 +92,17 @@ func (service *IPPoolService) buildIPSubnetIntentPath(IPPool *v1alpha2.IPPool, s
...
+	}
+
 	if IPPool.Spec.Type == common.IPPoolTypePrivate {
-		VPCInfo := commonctl.ServiceMediator.ListVPCInfo(IPPool.Namespace)
-		if len(VPCInfo) == 0 {
-			return nil
-		}
-		// TODO: Get the IPBlockPath by IPPool's namespace, external and private
-		IpBlockPath = String(fmt.Sprintf("/orgs/%s/projects/%s/infra/ip-blocks/block-test", VPCInfo[0].OrgID, VPCInfo[0].ProjectID))
+		IpBlockPath = String(VPCInfo[0].PrivateIpv4Blocks[0])
```
Comment (dantingl) [inline pkg/nsx/services/ippool/builder.go:103]: For IP exhausted case, we need to return an available ip block instead of the first one

Comment (lxiaopei) [reply pkg/nsx/services/ippool/builder.go:103]: agree

Comment (zhengxiexie) [reply pkg/nsx/services/ippool/builder.go:103]: would fix it after tao's patch merged.

File: pkg/nsx/services/ippool/builder.go [hunk 4]
```diff
@@ -92,15 +92,17 @@ func (service *IPPoolService) buildIPSubnetIntentPath(IPPool *v1alpha2.IPPool, s
...
+	} else {
+		IpBlockPath = String(VPCInfo[0].ExternalIpv4Blocks[0])
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/nsx/services/ippool/builder.go:105]: would fix it after tao's patch merged.

File: pkg/nsx/services/ippool/builder.go (hunk 3 above)
Comment (dantingl) [reply pkg/nsx/services/ippool/builder.go:103]: Add TODO for this one

File: pkg/nsx/services/ippool/builder.go [hunk 5]
```diff
@@ -92,15 +92,17 @@ func (service *IPPoolService) buildIPSubnetIntentPath(IPPool *v1alpha2.IPPool, s
 }
 
 func (service *IPPoolService) buildIPSubnet(IPPool *v1alpha2.IPPool, subnetRequest v1alpha2.SubnetRequest) *model.IpAddressPoolBlockSubnet {
-	// TODO: Get the IPBlockPath by IPPool's namespace, external and private
```
Comment (dantingl) [inline pkg/nsx/services/ippool/builder.go:95]: In line 94, what is SubnetRequest from v1alpha2?

Comment (zhengxiexie) [reply pkg/nsx/services/ippool/builder.go:95]: `// SubnetRequest defines the subnet allocation request.
type SubnetRequest struct {
	// PrefixLength defines prefix length for this subnet.
	PrefixLength int `json:"prefixLength,omitempty"`

	// IPFamily defines the IP family type for this subnet, could be IPv4 or IPv6.
	// This is optional, the default is IPv4.
	// +kubebuilder:validation:Enum=IPv4;IPv6
	// +kubebuilder:default=IPv4
	IPFamily string `json:"ipFamily,omitempty"`

	// Name defines the name of this subnet.
	Name string `json:"name"`
}`

**Comment by timdengyun** [comment]
please don't merge tao's patch into your patch: Dump error return from SDK to ApiError,
you can use it for your test, but should not be merged in your patch.

Also, please squash 3 commits into one commit.

File: pkg/controllers/ippool/ippool_controller.go [hunk 6]
```diff
@@ -134,10 +137,46 @@ func (r *IPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+			vpcNetworkConfig := common.ServiceMediator.GetVPCNetworkConfigByNamespace(obj.Namespace)
+			if vpcNetworkConfig == nil {
+				err := fmt.Errorf("operate failed: cannot get configuration for IPPool CR")
+				log.Error(nil, "failed to find VPCNetworkConfig for IPPool CR", "ippool", req.NamespacedName, "namespace %s", obj.Namespace)
```
Comment (TaoZou1) [inline pkg/controllers/ippool/ippool_controller.go:144]: why use log.Error but error is nil?

File: pkg/nsx/services/ippool/ippool.go [hunk 7]
```diff
@@ -71,6 +72,11 @@ func InitializeIPPool(service common.Service) (*IPPoolService, error) {
 
 func (service *IPPoolService) CreateOrUpdateIPPool(obj *v1alpha2.IPPool) (bool, bool, error) {
 	nsxIPPool, nsxIPSubnets := service.BuildIPPool(obj)
+	for _, ipSubnet := range nsxIPSubnets {
+		if *ipSubnet.IpBlockPath == "" {
+			return false, false, util.IPBlockAllExhaustedError{Desc: "all ip blocks are exhausted"}
+		}
```
Comment (TaoZou1) [inline pkg/nsx/services/ippool/ippool.go:78]: if first  *ipSubnet.IpBlockPath == "" , does that means all ipblock are exhausted? 
should we iterate all ipsubnet?

File: pkg/controllers/ippool/ippool_controller.go (hunk 6 above)
Comment (dantingl) [reply pkg/controllers/ippool/ippool_controller.go:144]: This is the way when we do not have error defined but want to print error message.
But I also saw Xie defined err in line 143, so I guess we can just leverage it?

File: pkg/controllers/ippool/ippool_controller.go [hunk 8]
```diff
@@ -134,10 +137,46 @@ func (r *IPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+			obj.Spec.Type = vpcNetworkConfig.DefaultSubnetAccessMode
+		}
+
 		subnetCidrUpdated, ipPoolSubnetsUpdated, err := r.Service.CreateOrUpdateIPPool(obj)
+		// check if ipblock is exhausted
+		apiErr, _ := util.DumpAPIError(err)
+		if apiErr != nil {
+			for _, apiErrItem := range apiErr.RelatedErrors {
+				if *apiErrItem.ErrorCode == 520012 {
+					if strings.Contains(*apiErrItem.ErrorMessage, "does not have spare capacity to satisfy") {
```
Comment (dantingl) [inline pkg/controllers/ippool/ippool_controller.go:157]: Let's do not compare error message, NSX may change error message, error code is good enough

File: pkg/controllers/ippool/ippool_controller.go (hunk 6 above)
Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:144]: ok

File: pkg/nsx/services/ippool/ippool.go (hunk 7 above)
Comment (zhengxiexie) [reply pkg/nsx/services/ippool/ippool.go:78]: no need, if *ipSubnet.IpBlockPath == "", it represents all exhausted, it is the logic indeed.

File: pkg/controllers/ippool/ippool_controller.go (hunk 8 above)
Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:157]: I'm not sure 520012 is only for exhausted, so add an error message check to insure more safety.

Comment (dantingl) [reply pkg/controllers/ippool/ippool_controller.go:157]: Yes. 520012 is only for exhaustion case

Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:157]: OK

File: pkg/controllers/ippool/ippool_controller.go [hunk 9]
```diff
@@ -134,10 +136,45 @@ func (r *IPPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+				}
+			}
+		}
+
 		if err != nil {
-			log.Error(err, "operate failed, would retry exponentially", "ippool", req.NamespacedName)
 			updateFail(r, &ctx, obj, &err)
+			// if all ip blocks are exhausted, we should not retry
+			if errors.As(err, &util.IPBlockAllExhaustedError{}) {
+				return resultNormal, nil
```
Comment (TaoZou1) [inline pkg/controllers/ippool/ippool_controller.go:175]: add a log.Error before turn?

Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:176]: ok

**Review by TaoZou1 — APPROVED** [review]

**Pull request merged**